linalg/hypre.cpp

```diff
@@ -3092,6 +3180,75 @@ HypreBoomerAMG::HypreBoomerAMG(HypreParMatrix &A) : HypreSolver(&A)
    SetDefaultOptions();
 }
 
+void HypreBoomerAMG::Mult(const HypreParVector &b, HypreParVector &x) const
```

Comment:
Is this necessary when we have `HypreSolver::Mult`? Is this needed because of the timings? Can it be made more general for all hypre solvers?

Reply:
Needed for timings. Must includes the construction of the hypre objects (CG, GMRES, BoomerAMG, etc.), so it's not particularly natural to have a general mult unless you abstract away the setup functions

Reply:
I think `HypreSolver::Mult` might have what you need (in particular, it has abstract `SetupFcn` and `SolveFcn` for setup and solve phases)

Reply:
Okay, that could work. I guess compressing all Hypre objects to a single Mult function is a bit out of scope of what I was trying to do here, I just thought BoomerAMG should also have timing available, similar to GMRES and CG.

Reply:
The reason I mention it is because all the `Mult` functions were previously "compressed" — there was only `HypreSolver::Mult`, and that was used for all of the subclasses.

This PR introduced a new special `Mult` just for `HypreBoomerAMG`. But it seems like instead of making a new specialized `Mult`, the generic `HypreSolver::Mult` could be modified instead.

Reply:
? There are HypreGMRES::Mult and HyprePCG::Mult, which are the functions I largely copied the structure of to make HypreBoomerAMG::Mult

Reply:
Fair enough, those two specializations exist. All the remaining subclasses of HypreSolver (AMS, ADS, ParaSails, Euclid, etc.) fall back on `HypreSolver::Mult`.

Generally speaking, I'm not a fan of having the exact same (nontrivial) logic repeated three or four times (with very minor — perhaps accidental? — modifications between them). Already the CG and GMRES implementations are somewhat divergent: the CG version has `b.HostRead(); x.HostReadWrite();`, and the GMRES version doesn't. The `print_level` is also not treated uniformly in the two solvers.

I agree that unifying the CG and GMRES version of Mult is beyond the scope of this PR, but I am concerned that adding an additional copy & paste Mult specialization just increases the maintenance burden.

Reply:
Sure, I see the benefit of a uniform Mult. The Mult can be removed if it must, I will just add it to my own branches where I need it, or the timing could be added to the general Mult, but then the print output would not say, e.g., "CG iterations." I also think the print_level output should be consistent between the hypre solvers and MFEM solvers, even if it is inconsistent with internal hypre. That's always confused me a bit, having to change between print_levels 1 or 2 depending on if I am using MFEM or Hypre solvers.

Reply:
One solution for now would be to get rid of `HypreBoomerAMG::Mult` from the current PR, but also make an issue on this subject (and link back to this discussion). That way we can eventually make a separate PR to try to tackle the code duplication/inconsistencies.